File: python/day06.py
import re, itertools, functools, operator

ops = {'+': operator.add, '*': operator.mul}

def parse_ops(line):
    return [tok for tok in re.sub(r"\s+", " ", line).strip().split(" ") if tok]

def parse_grid_tokens(lines):
    # Split each row into tokens (numbers), then transpose into columns
    rows = [r.split() for r in lines]
    width = max(len(r) for r in rows)
    for r in rows:
        if len(r) != width:
            raise ValueError("Ragged rows in token grid")
    cols = list(map(list, zip(*rows)))  # transpose
    return [[int(x) for x in col] for col in cols]

def parse_grid_constructed(lines):
    # Character-wise transpose; group contiguous digit-columns into numbers
    maxlen = max(len(r) for r in lines)
    padded = [r.ljust(maxlen) for r in lines]
    transposed = ["".join(row[c] for row in padded) for c in range(maxlen)]
    groups = [
        list(element)
        for key, element in itertools.groupby(
            transposed, lambda x: not re.match(r"^\s+$", x)
        )
        if key
    ]
    return [[int(x) for x in group] for group in groups]

def eval_and_print(label, operators, columns):
    total = 0
    print(f"\n--- {label} ---")
    for op, nums in zip(operators, columns):
        res = functools.reduce(ops[op], nums)
        print(f"{op} on {nums} -> {res}")
        total += res
    print(f"{label} Total = {total}")
    return total

def solve(filename="input/day06.txt"):
    with open(filename) as f:
        data = [line.rstrip("\n") for line in f if line.strip()]

    op_line = data[-1]
    body = data[:-1]
    operators = parse_ops(op_line)

    # L2R: simple arithmetic
    l2r_cols = parse_grid_tokens(body)
    total_l2r = eval_and_print("Left-to-Right", operators, l2r_cols)

    # R2L: constructed numbers
    r2l_cols = parse_grid_constructed(body)[::-1]
    total_r2l = eval_and_print("Right-to-Left", operators[::-1], r2l_cols)

    return total_l2r, total_r2l

if __name__ == "__main__":
    l2r, r2l = solve("input/day06.txt")
    print(f"\nDay 6 Answer (Left-to-Right): {l2r}")
    print(f"Day 6 Answer (Right-to-Left): {r2l}")

File: perl/day06.pl
#!/usr/bin/perl
use strict;
use warnings;

sub parse_ops {
    my ($line) = @_;
    $line =~ s/\s+/ /g;
    $line =~ s/^\s+|\s+$//g;
    my @ops = grep { $_ ne '' } split(/ /, $line);
    return \@ops;
}

sub parse_grid_tokens {
    my ($lines) = @_;
    my @rows = map { [ split(/\s+/, $_) ] } @$lines;
    my $width = 0;
    for my $r (@rows) {
        $width = @$r if @$r > $width;
    }
    # transpose
    my @cols;
    for my $c (0 .. $width-1) {
        my @col;
        for my $r (@rows) {
            push @col, $r->[$c] if defined $r->[$c];
        }
        push @cols, [ map { int($_) } @col ];
    }
    return \@cols;
}

sub parse_grid_constructed {
    my ($lines) = @_;
    my $maxlen = 0;
    $maxlen = length($_) > $maxlen ? length($_) : $maxlen for @$lines;
    my @padded = map { $_ . (' ' x ($maxlen - length($_))) } @$lines;

    my @transposed;
    for my $c (0 .. $maxlen-1) {
        my $s = join('', map { substr($_, $c, 1) } @padded);
        push @transposed, $s;
    }

    # group contiguous digit columns
    my @groups;
    my $current = [];
    for my $col (@transposed) {
        if ($col =~ /\d/) {
            push @$current, $col;
        } else {
            if (@$current) {
                push @groups, [ map { int($_) } @$current ];
                $current = [];
            }
        }
    }
    if (@$current) {
        push @groups, [ map { int($_) } @$current ];
    }
    return \@groups;
}

sub eval_and_print {
    my ($label, $ops, $cols) = @_;
    my $total = 0;
    print "\n--- $label ---\n";
    for my $i (0 .. $#$ops) {
        my $op = $ops->[$i];
        my @nums = @{ $cols->[$i] };
        my $res;
        if ($op eq '+') {
            $res = 0; $res += $_ for @nums;
        } elsif ($op eq '*') {
            $res = 1; $res *= $_ for @nums;
        }
        print "$op on [@nums] -> $res\n";
        $total += $res;
    }
    print "$label Total = $total\n";
    return $total;
}

sub solve {
    my ($filename) = @_;
    open my $fh, '<', $filename or die "Cannot open $filename: $!";
    my @data = grep { /\S/ } map { chomp; $_ } <$fh>;
    close $fh;

    my $op_line = $data[-1];
    my @body = @data[0 .. $#data-1];
    my $ops = parse_ops($op_line);

    # L2R
    my $l2r_cols = parse_grid_tokens(\@body);
    my $total_l2r = eval_and_print("Left-to-Right", $ops, $l2r_cols);

    # R2L
    my $r2l_cols = [ reverse @{ parse_grid_constructed(\@body) } ];
    my $rev_ops  = [ reverse @$ops ];
    my $total_r2l = eval_and_print("Right-to-Left", $rev_ops, $r2l_cols);

    return ($total_l2r, $total_r2l);
}

my ($l2r, $r2l) = solve("input/day06.txt");
print "\nDay 6 Answer (Left-to-Right): $l2r\n";
print "Day 6 Answer (Right-to-Left): $r2l\n";
